src/Examples/Tools/DegeneracyHunter/degeneracy_hunter.ipynb

```diff
@@ -0,0 +1,1538 @@
+      "Dual infeasibility......:   1.5188693260016487e-13    1.5188693260016487e-13\n",
+      "Constraint violation....:   9.9976108502985994e-09    9.9976108502985994e-09\n",
+      "Complementarity.........:   9.2217032157601989e-10    9.2217032157601989e-10\n",
+      "Overall NLP error.......:   9.9976108502985994e-09    9.9976108502985994e-09\n",
+      "\n",
+      "\n",
+      "Number of objective function evaluations             = 111\n",
+      "Number of objective gradient evaluations             = 17\n",
+      "Number of equality constraint evaluations            = 111\n",
+      "Number of inequality constraint evaluations          = 111\n",
+      "Number of equality constraint Jacobian evaluations   = 17\n",
+      "Number of inequality constraint Jacobian evaluations = 17\n",
+      "Number of Lagrangian Hessian evaluations             = 16\n",
+      "Total CPU secs in IPOPT (w/o function evaluations)   =      0.006\n",
+      "Total CPU secs in NLP function evaluations           =      0.000\n",
+      "\n",
+      "EXIT: Optimal Solution Found.\n",
+      "x[ 1 ]= 1.0000000099975996\n",
+      "x[ 2 ]= 0.0\n",
+      "x[ 3 ]= 0.0\n"
+     ]
+    }
+   ],
+   "source": [
+    "opt.options['max_iter'] = 50\n",
+    "opt.solve(m2, tee=True)\n",
+    "\n",
+    "for i in m2.I:\n",
+    "    print(\"x[\",i,\"]=\",m2.x[i]())"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We got luck here. Ipopt implements several algorithmic and numerical safeguards to handle (mildy) degenerate equations. Nevertheless, notice the last column of the Ipopt output labeled `ls`. This is the number of linesearch evaluations. For iterations 0 to 11, `ls` is 1, which means Ipopt is taking full steps. For iterations 12 to 16, however, `ls` is greater than 20. This means Ipopt is struggling (a little) to converge to the solution."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Check the rank of the Jacobian of the equality constraints"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "\n",
+      "Checking rank of Jacobian of equality constraints...\n",
+      "Model contains 2 equality constraints and 3 variables.\n",
+      "Computing the 1 smallest singular value(s)\n",
+      "Smallest singular value(s):\n",
+      "0.000E+00\n"
+     ]
+    }
+   ],
+   "source": [
+    "n_deficient = dh2.check_rank_equality_constraints()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "A singular value near 0 indicates the Jacobian of the equality constraints is rank deficient. For each near-zero signular value, there is likely one degenerate constraint."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Identify candidate degenerate constraints"
+   ]
+  },
+  {
```

Comment:
It is not clear to me what the candidate equations found are (the method does not return an object/result to the notebook), or how they will be used. It would be helpful to explain what this step is doing in the background and how will will use that result.

E.g. this method does X to find Y which is stored in the dh2 object. We will use Y in the next step to do Z.

Reply:
Good suggestion. I added a few sentences.